### DISCUSSION TITLE
Adding ServiceTime and Demands is not yielding results for big dataset(1900 stores)).

### DISCUSSION THREAD

**Discussion by SruthyJohn** [google/or-tools#3242] [Routing (and legacy CP) questions]
Hi Team,
The below Routing solver is working fine and I am getting results upto 200 stores and 5 vans.
But as per my business scenario, when i tried with 1873 stores and 33vans with servicetime for stores, demands, max store count for a vehicle constraints I am getting the solution as null.

**DataSet:**
Stores - 1873 
TimeWindow for stores - 28800 - 64800 
ServiceTime for each store - 900 (15 mins)
Demand for each store - 118 (average)

Vans - 33
Capacity - 100000
MaxStore Count - 60

Solution =null (When adding Demand, ServiceTime of 900secs for each store)
Solution = value (TimeWindow, Max StoreCount constraints alone)

Please find the attached gist for the program.
https://gist.github.com/SruthyJohn/72aab8773249eb4584cd074cc5239a9c

**Comment by vuongdanghuy** [comment]
I think you should check the status of routing model after solving to see what's happened. solution is null normally because of routing timeout or routing model is not correct.
Reference:
https://github.com/google/or-tools/blob/49b6301e1e1e231d654d79b6032e79809868a70e/ortools/constraint_solver/routing.cc#L3352-L3360

**Reply by SruthyJohn** [reply]
> I think you should check the status of routing model after solving to see what's happened. solution is null normally because of routing timeout or routing model is not correct. Reference:
> 
> https://github.com/google/or-tools/blob/49b6301e1e1e231d654d79b6032e79809868a70e/ortools/constraint_solver/routing.cc#L3352-L3360

Thanks [vuongdanghuy](https://github.com/vuongdanghuy) for your reply.
I am getting the ROUTING_FAIL_TIMEOUT as status. I tried increasing the time, but still getting the same. And also, the same dataset working perfectly if i comment out the addition of ServiceTime in the timeCallbackIndex  and return results within 5 mins.

```cs
int timeCallbackIndex = routing.RegisterTransitCallback((long fromIndex, long toIndex) =>
                    {
                        var fromNode = manager.IndexToNode(fromIndex);
                        var toNode = manager.IndexToNode(toIndex);
                        return data.TimeMatrix[fromNode, toNode]   **//+ data.ServiceTimes[fromNode]);** 
                    });
```

**Reply by Mizux** [reply]
your problem maybe infeasible or the solver have hard time to find a first solution.
You may try to allow the drop of some nodes % an high penalty cost...

see: https://developers.google.com/optimization/routing/penalties

**Reply by SruthyJohn** [reply]
> your problem maybe infeasible or the solver have hard time to find a first solution. You may try to allow the drop of some nodes % an high penalty cost...
> 
> see: https://developers.google.com/optimization/routing/penalties

Hi Mizux, I tried adding disjunction. But then it is dropping all nodes.

**Reply by Mizux** [reply]
Does your ServiceTimes an integer array ? Did you check the values returned ?

**Reply by vuongdanghuy** [reply]
> > your problem maybe infeasible or the solver have hard time to find a first solution. You may try to allow the drop of some nodes % an high penalty cost...
> > see: https://developers.google.com/optimization/routing/penalties
> 
> Hi Mizux, I tried adding disjunction. But then it is dropping all nodes.

You should check ServiceTimes as Mizux suggested to see if it's correct.
I see you're creating a time dimension for each vehicle using a for loop, but it'll only create one dimension with name 'Time' in the first iteration with max capacity equal to data.VehicleMaxDurations[0]. So maybe the time capacity of each vehicle is too small or node time window constraint can't be satisfied after adding service times. You can check it by examining the time matrix.
If it's still not working, try debug it by using small dataset with service time, may be 20-30 nodes, 5-10 vehicles to check model correctness.

**Reply by SruthyJohn** [reply]
Hi @Mizux,
ServiceTimes is long array. Checked the values coming and it is as expected.
How will I check the data will jot have the solution and alert user there is some error in the passed data, so that he can rectify and resend to the or-tools to get a solution or as developers we can use it for log purposes? 

Hi @vuongdanghuy ,
I checked with adding single Time dimension as below,

`routing.AddDimensionWithVehicleCapacity(timeCallbackIndex,0, data.VehicleMaxDurations,false,"Time");`

But this too is not helping, solution is null and status is ROUTING_FAIL_TIMEOUT. Any idea on how will we validate the data even before passing to OR-tools?

**Reply by vuongdanghuy** [reply]
> Hi @Mizux, ServiceTimes is long array. Checked the values coming and it is as expected. How will I check the data will jot have the solution and alert user there is some error in the passed data, so that he can rectify and resend to the or-tools to get a solution or as developers we can use it for log purposes?
> 
> Hi @vuongdanghuy , I checked with adding single Time dimension as below,
> 
> `routing.AddDimensionWithVehicleCapacity(timeCallbackIndex,0, data.VehicleMaxDurations,false,"Time");`
> 
> But this too is not helping, solution is null and status is ROUTING_FAIL_TIMEOUT. Any idea on how will we validate the data even before passing to OR-tools?

Did you try running your code with smaller dataset? It's better to debug code with fewer number of nodes and vehicles. Let's try for example pick randomly 10 nodes and 2 or 3 vehicles from big dataset with ServiceTime. If it run well, then try a little bigger, says 50 nodes, 5-6 vehicles with ServiceTime and see if it still return null

**Reply by Mizux** [reply]
1. You should add some slacks otherwise your route are nearly impossible.
i.e. you are asking to the solver to find a route with zero waiting time between visit aka kind of perfect busy route.
see: https://gist.github.com/SruthyJohn/72aab8773249eb4584cd074cc5239a9c#file-program-cs-L56

2. You should also allow to drop nodes with very high penalty so solver can at least find a partial solution then you rely on the gls to fix it.
i.e. some time (nearly always) it's better to let the solver find a partial solution then rely on the guided local search for letting the solver find a solution with the previously dropped node.

3. You may use `searchParameters.LogSearch = true;` to see what the solver is doing.
note: you can follow the solver GLS and if penalty for dropping node is the dominant factor you can spot along the search how many nodes are dropped looking at the cirrent objective significant digits.
e.g. if your penalty for dropping node is 1,000,000 while any arc cost is around few hundred an objective of 42,00X,YZW means you likely have 42 dropping nodes...
note: But here it seems the solver just can't find a first solution.
